library/core/resource-loader/src/main/java/org/quiltmc/qsl/resource/loader/impl/ModNioResourcePack.java

```diff
@@ -0,0 +1,215 @@
+	private Set<String> namespaceCache;
+
+	public ModNioResourcePack(ModMetadata modInfo, Path path, ResourceType type, AutoCloseable closer, ResourcePackActivationType activationType) {
+		super(null);
+		this.modInfo = modInfo;
+		this.basePath = path.toAbsolutePath().normalize();
+		this.type = type;
+		this.cacheable = modInfo.getId().equals("minecraft");
+		this.closer = closer;
+		this.separator = basePath.getFileSystem().getSeparator();
+		this.activationType = activationType;
+	}
+
+	private Path getPath(String filename) {
+		Path childPath = basePath.resolve(filename.replace("/", separator)).toAbsolutePath().normalize();
+
+		if (childPath.startsWith(basePath) && Files.exists(childPath)) {
+			return childPath;
+		} else {
+			return null;
+		}
+	}
+
+	@Override
+	protected InputStream openFile(String filename) throws IOException {
+		InputStream stream;
+
+		Path path = getPath(filename);
+
+		if (path != null && Files.isRegularFile(path)) {
+			return Files.newInputStream(path);
+		}
+
+		stream = ModResourcePackUtil.openDefault(this.modInfo, this.type, filename);
+
+		if (stream != null) {
+			return stream;
+		}
+
+		// ReloadableResourceManagerImpl gets away with FileNotFoundException.
+		throw new FileNotFoundException("\"" + filename + "\" in Fabric mod \"" + modInfo.getId() + "\"");
+	}
+
+	@Override
+	protected boolean containsFile(String filename) {
+		if (ModResourcePackUtil.containsDefault(modInfo, filename)) {
+			return true;
+		}
+
+		Path path = this.getPath(filename);
+		return path != null && Files.isRegularFile(path);
+	}
+
+	@Override
+	public Collection<Identifier> findResources(ResourceType type, String namespace, String path, int depth, Predicate<String> predicate) {
+		var ids = new ArrayList<Identifier>();
+		String nioPath = path.replace("/", separator);
+
+		Path namespacePath = this.getPath(type.getDirectory() + "/" + namespace);
+
+		if (namespacePath != null) {
+			Path searchPath = namespacePath.resolve(nioPath).toAbsolutePath().normalize();
+
+			if (Files.exists(searchPath)) {
+				try {
+					Files.walk(searchPath, depth)
+							.filter(Files::isRegularFile)
+							.filter((p) -> {
+								String filename = p.getFileName().toString();
+								return !filename.endsWith(".mcmeta") && predicate.test(filename);
+							})
+							.map(namespacePath::relativize)
+							.map((p) -> p.toString().replace(separator, "/"))
+							.forEach((s) -> {
+								try {
+									ids.add(new Identifier(namespace, s));
+								} catch (InvalidIdentifierException e) {
+									LOGGER.error(e.getMessage());
+								}
+							});
+				} catch (IOException e) {
+					LOGGER.warn("findResources at " + path + " in namespace " + namespace + ", mod " + modInfo.getId() + " failed!", e);
+				}
+			}
+		}
+
+		return ids;
+	}
+
+	protected void warnInvalidNamespace(String s) {
+		LOGGER.warn("Fabric NioResourcePack: ignored invalid namespace: {} in mod ID {}", s, modInfo.getId());
```

Comment:
Fabric? Don't you mean Quilt?
```suggestion
		LOGGER.warn("Quilt NioResourcePack: ignored invalid namespace: {} in mod ID {}", s, modInfo.getId());
```